### PR TITLE
Implement and Document new release workflow. 

### PR DESCRIPTION
**Pull request by csala**
Implement and Document the new release workflow.

Also fix/cleanup the installation requirements, and tests config, and integrate with codecov.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -1,40 +1,22 @@
...
   - 3.4
   - 3.5
   - 3.6
-
-matrix:
-  include:
-    - python: 3.7
-      dist: xenial
-      sudo: true
+  - 3.7
```
Comment (micahjsmith) [inline .travis.yml]: why this change? to the best of my understanding, the previous state was the correct way to test python 3.7: https://github.com/travis-ci/travis-ci/issues/9815. the way you have done it seems like it should not work because sudo is required. and removing the matrix include entry means that you would have to enable sudo for all environments which is less than ideal.

File: .travis.yml [hunk 2]
```diff
@@ -49,4 +31,4 @@ deploy:
     target-branch: gh-pages
     on:
       branch: master
-      python: 3.7
+      python: 3.6
```
Comment (micahjsmith) [inline .travis.yml]: why this change?

File: Makefile [hunk 3]
```diff
@@ -26,77 +25,171 @@ export PRINT_HELP_PYSCRIPT
...
+	pip install -e .[dev]
+
+
+# LINT TARGETS
+
+.PHONY: lint
 lint: ## check style with flake8 and isort
-	flake8 btb tests examples
-	isort -c --recursive btb tests examples
+	flake8 btb tests
```
Comment (micahjsmith) [inline Makefile]: why this change? examples should still be linted.

File: Makefile [hunk 4]
```diff
@@ -26,77 +25,171 @@ export PRINT_HELP_PYSCRIPT
...
+	isort -c --recursive btb tests
 
+.PHONY: fix-lint
 fix-lint: ## fix lint issues using autoflake, autopep8, and isort
-	autopep8 --in-place --recursive --aggressive --aggressive btb tests examples
-	isort --apply --atomic --recursive btb tests examples
+	find btb -name '*.py' | xargs autoflake --in-place --remove-all-unused-imports --remove-unused-variables
```
Comment (micahjsmith) [inline Makefile]: I suggest abandoning autoflake.

and there is no need to re-separate the autopep8 and isort commands, they can be a single command that applies to `btb tests examples`.

File: Makefile [hunk 5]
```diff
@@ -26,77 +25,171 @@ export PRINT_HELP_PYSCRIPT
...
-	rm -f docs/api/*.rst
-	$(MAKE) -C docs clean
 
+# DOCS TARGETS
+
+.PHONY: docs
 docs: clean-docs ## generate Sphinx HTML documentation, including API docs
-	sphinx-apidoc --module-first --separate -o docs/api/ btb
 	$(MAKE) -C docs html
+	touch docs/_build/html/.nojekyll
```
Comment (micahjsmith) [inline Makefile]: not needed. we already rely on `sphinx.ext.githubpages` which creates the `.nojekyll` file.

File: setup.cfg [hunk 6]
```diff
@@ -31,7 +31,6 @@ ignore = W503,W504
 
 [isort]
 include_trailing_comment = True
-known_first_party = btb
```
Comment (micahjsmith) [inline setup.cfg]: I don't have any objection to remove it, but I when we added it, it was for a good reason.

File: .travis.yml (hunk 1 above)
Comment (csala) [reply .travis.yml]: The change is because this configuration is simpler and it is proved to work.

File: .travis.yml (hunk 2 above)
Comment (csala) [reply .travis.yml]: Because there is no advantage on one over the other, but this one is proved to work with the upper configuration, while 3.7 hasn't.

File: Makefile (hunk 3 above)
Comment (csala) [reply Makefile]: because I forgot to add them. Adding them back.

File: Makefile (hunk 4 above)
Comment (csala) [reply Makefile]: The need of this change is simply that this is what we have in the cookiecutter-pypackage.
We can always modify it in the future when we change the cookiectter template and port the changes here.

File: Makefile (hunk 5 above)
Comment (csala) [reply Makefile]: removing

File: README.md [hunk 7]
```diff
@@ -7,12 +7,20 @@
 
 A simple, extensible backend for developing auto-tuning systems.
 
-[![PyPI Shield](https://img.shields.io/pypi/v/baytune.svg)](https://pypi.python.org/pypi/baytune)
-[![Travis CI Shield](https://travis-ci.org/HDI-Project/BTB.svg?branch=master)](https://travis-ci.org/HDI-Project/BTB)
+[![PyPi][pypi-img]][pypi-url]
```
[the remark this answers is not in the document]
Comment (csala) [reply README.md]: I'll change it.

File: setup.cfg (hunk 6 above)
Comment (csala) [reply setup.cfg]: The was a reason and it was a good reason: it failed otherwise.
But the reason of the failure was that the tox.ini configuration was not right, as the project was not being installed in the flake8 environment.
Now the tox.ini config has been fixed, so this can be removed.

File: .travis.yml (hunk 1 above)
Comment (micahjsmith) [reply .travis.yml]: where? the best reference I have seen is the linked travis issue.

Comment (csala) [reply .travis.yml]: Originally I got it from here: https://github.com/travis-ci/travis-ci/issues/9831
Later on I made it work here: https://github.com/csala/smapy/blob/master/.travis.yml
Also the travis python guide points at a similar approach: https://docs.travis-ci.com/user/languages/python/

File: .travis.yml (hunk 2 above)
Comment (csala) [reply .travis.yml]: I realized I ended up upgrading to 3.7 in the other project where I had used this configuration, so I reverted it to 3.7 also here.

File: .travis.yml (hunk 1 above)
Comment (micahjsmith) [reply .travis.yml]: fair enough
